### PR TITLE
add logging library to EventListener

### PR DESCRIPTION
**Pull request by akihikokuroda**
# Changes

This PR for the issue #76

The logging config is in config-logging.yaml file. The log component string is eventlistener and the default setting is info. The Logger is set in the sink.Resource so that any member function of the struct can use the logger.

Note: The config-logging-triggers ConfigMap is created in the namespace where the eventlistener is created when it doesn't exist.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior

```

### CODE REVIEW

File: docs/eventlisteners.md [hunk 1]
```diff
@@ -14,6 +14,10 @@ does what has been specified in the corresponding
 services' via their cluster DNS. For external services to connect to your
 cluster (e.g. GitHub sending webhooks), check out the guide on [exposing eventlisteners](./exposing-eventlisteners.md)
 
+When the `EventListener` is created in the different namespace from the trigger component, `config-logging-triggers` ConfigMap
+is created for the logging configuration in the namespace when it doesn't exist..  The ConfigMap with the defatul configuration can be created
```
Comment (ncskier) [inline docs/eventlisteners.md]: typos:
`..` -> `.`
`defatul` -> `default`

File: pkg/reconciler/v1alpha1/eventlistener/eventlistener.go [hunk 2]
```diff
@@ -204,7 +204,60 @@ func (c *Reconciler) reconcileService(el *v1alpha1.EventListener) error {
...
+			c.Logger.Fatalf("Failed to create logging config: %s", err)
+			return err
+		}
+	}
+	return nil
+}
+
 func (c *Reconciler) reconcileDeployment(el *v1alpha1.EventListener) error {
+	// check logging config, create if it doesn't exist
+	err := c.createLoggingConfig(el)
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply pkg/reconciler/v1alpha1/eventlistener/eventlistener.go]: I'll change it to `reconcileLoggingConfig`.

File: pkg/reconciler/v1alpha1/eventlistener/eventlistener.go [hunk 3]
```diff
@@ -204,7 +204,60 @@ func (c *Reconciler) reconcileService(el *v1alpha1.EventListener) error {
 	return nil
 }
 
+func (c *Reconciler) createLoggingConfig(el *v1alpha1.EventListener) error {
+	_, err := c.KubeClientSet.CoreV1().ConfigMaps(el.Namespace).Get("config-logging-triggers", metav1.GetOptions{})
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply pkg/reconciler/v1alpha1/eventlistener/eventlistener.go]: I didn't know the function.  I change to use it.

File: pkg/reconciler/v1alpha1/eventlistener/eventlistener.go [hunk 4]
```diff
@@ -204,7 +204,60 @@ func (c *Reconciler) reconcileService(el *v1alpha1.EventListener) error {
...
+						"\"lineEnding\": \"\"," +
+						"\"levelEncoder\": \"\"," +
+						"\"timeEncoder\": \"\"," +
+						"\"durationEncoder\": \"\"," +
+						"\"callerEncoder\": \"\"" +
+						"}" +
+						"}",
+				},
+			},
+		)
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply pkg/reconciler/v1alpha1/eventlistener/eventlistener.go]: I'll take them out.

File: docs/eventlisteners.md (hunk 1 above)
Comment (akihikokuroda) [reply docs/eventlisteners.md]: I fix them.  Thanks!

**Comment by ncskier** [comment]
Can you please squash your commits into one. Also, can you please provide a little more detail in your commit message. The description in this PR is great, so you can use something like that for the commit message to follow the Tekton [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages) 👍

**Comment by akihikokuroda** [comment]
Thanks for review.  I updated the commit message and squashed them into 1.

**Comment by ncskier** [comment]
I just tested this on OpenShift, which has stricter RBAC than Docker for Desktop, and I got an error in the EventListener logs about not having permission to `list` ConfigMaps:
```
k8s.io/client-go/informers/factory.go:131: Failed to list *v1.ConfigMap: configmaps is forbidden: User "system:serviceaccount:demo:tekton-triggers-example-sa" cannot list resource "configmaps" in API group "" in the namespace "demo"
```
I think that we need to document that the EventListener role must have
```
- apiGroups: [""]
  resources: ["configmaps"]
  verbs: ["list", "watch"]
```
in order to set up the logger. We should also add this to the [role.yaml](https://github.com/tektoncd/triggers/blob/master/examples/role-resources/role.yaml) file and any other examples we have.

Also, is there a plan for cleaning up these ConfigMaps after an EventListener is deleted? I'm not sure if this is something we need to be worried about, but it would probably be nice if we deleted the ConfigMap when there are no EventListeners in the namespace.

**Comment by akihikokuroda** [comment]
I updated [role.yaml](https://github.com/tektoncd/triggers/blob/master/examples/role-resources/role.yaml), [eventlistener.md](https://github.com/tektoncd/triggers/blob/master/docs/eventlisteners.md) and [eventlistener_test.go](https://github.com/tektoncd/triggers/blob/master/test/eventlistener_test.go)

File: cmd/eventlistenersink/main.go [hunk 5]
```diff
@@ -18,23 +18,56 @@ package main
 
 import (
 	"fmt"
+	tktnlogging "github.com/tektoncd/triggers/pkg/logging"
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply cmd/eventlistenersink/main.go]: I change it.

File: cmd/webhook/main.go [hunk 6]
```diff
@@ -18,20 +18,16 @@ package main
 
 import (
 	"context"
-	"flag"
 	"log"
 
 	"github.com/tektoncd/pipeline/pkg/system"
 	"github.com/tektoncd/triggers/pkg/apis/triggers/v1alpha1"
-
+	tktnlogging "github.com/tektoncd/triggers/pkg/logging"
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply cmd/webhook/main.go]: same as above.

File: pkg/reconciler/v1alpha1/eventlistener/eventlistener.go [hunk 7]
```diff
@@ -204,7 +204,61 @@ func (c *Reconciler) reconcileService(el *v1alpha1.EventListener) error {
...
+						"\"stacktraceKey\": \"stacktrace\"," +
+						"\"lineEnding\": \"\"," +
+						"\"levelEncoder\": \"\"," +
+						"\"timeEncoder\": \"\"," +
+						"\"durationEncoder\": \"\"," +
+						"\"callerEncoder\": \"\"" +
+						"}" +
+						"}",
+				},
+			},
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply pkg/reconciler/v1alpha1/eventlistener/eventlistener.go]: I believe it requires clusterrole.  I don't think it's too much for the eventlistener.  What do you think?

File: pkg/reconciler/v1alpha1/eventlistener/eventlistener.go [hunk 8]
```diff
@@ -204,7 +204,61 @@ func (c *Reconciler) reconcileService(el *v1alpha1.EventListener) error {
...
+		)
+		if err != nil {
+			c.Logger.Fatalf("Failed to create logging config: %s", err)
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply pkg/reconciler/v1alpha1/eventlistener/eventlistener.go]: This error causes the eventlistener pod not started.  It is not fatal for the controller but for the eventlistener.  What do you think?

File: test/eventlistener_test.go [hunk 9]
```diff
@@ -221,6 +229,45 @@ func TestEventListenerCreate(t *testing.T) {
...
+					"\"durationEncoder\": \"\"," +
+					"\"callerEncoder\": \"\"" +
+					"}" +
+					"}",
+			},
+		},
+	)
+	if err != nil {
+		t.Fatalf("Failed to create EventListener: %s", err)
+	}
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply test/eventlistener_test.go]: I take it out.

File: examples/role-resources/role.yaml [hunk 10]
```diff
@@ -9,3 +9,6 @@ rules:
 - apiGroups: ["tekton.dev"]
   resources: ["pipelineruns", "pipelineresources", "taskruns"]
   verbs: ["create"]
+- apiGroups: [""]
+  resources: ["configmaps"]
+  verbs: ["get", "list", "watch"]
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply examples/role-resources/role.yaml]: I update.

File: docs/eventlisteners.md [hunk 11]
```diff
@@ -3,8 +3,27 @@
...
+- apiGroups: ["tekton.dev"]
+  resources: ["eventlisteners", "triggerbindings", "triggertemplates", "tasks", "taskruns"]
+  verbs: ["get"]
+- apiGroups: ["tekton.dev"]
+  resources: ["pipelineruns", "pipelineresources", "taskruns"]
+  verbs: ["create"]
+- apiGroups: [""]
+  resources: ["configmaps"]
+  verbs: ["get", "list", "watch"]
+```
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply docs/eventlisteners.md]: I do it.

File: pkg/reconciler/v1alpha1/eventlistener/eventlistener.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply pkg/reconciler/v1alpha1/eventlistener/eventlistener.go]: I change to "error".

File: cmd/eventlistenersink/main.go [hunk 12]
```diff
@@ -18,23 +18,56 @@ package main
...
+	"github.com/tektoncd/triggers/pkg/logging"
 	"github.com/tektoncd/triggers/pkg/sink"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"knative.dev/pkg/signals"
+)
+
+// EventListenerLogKey is the name of the logger for the eventlistener cmd
+const (
+	EventListenerLogKey = "eventlistener"
```
Comment (ncskier) [inline cmd/eventlistenersink/main.go]: Oops, I think I missed this before, but the comment should be directly above the `EventListenerLogKey` line.
```
// EventListenerLogKey is the name of the logger for the eventlistener cmd
EventListenerLogKey = "eventlistener"
```

File: pkg/reconciler/v1alpha1/eventlistener/eventlistener.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ncskier) [reply pkg/reconciler/v1alpha1/eventlistener/eventlistener.go]: Yeah, if it requires another rule to give the EventListener access to the `tekton-pipelines` namespace, then it probably isn't worth the change; thanks!

File: pkg/reconciler/v1alpha1/eventlistener/eventlistener.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ncskier) [reply pkg/reconciler/v1alpha1/eventlistener/eventlistener.go]: thanks!

**Comment by ncskier** [comment]
Thank you for the updates @akihikokuroda! 🙂 I left a comment about a small nit that I must've missed earlier, but everything looks good to merge after that. 👍

File: cmd/eventlistenersink/main.go (hunk 12 above)
Comment (akihikokuroda) [reply cmd/eventlistenersink/main.go]: oh, I fix it.

**Pull request merged**